### DISCUSSION TITLE
I have been using hall_sensor as my main input interface, create fake magnetic pin to use as pushbuttons

### DISCUSSION THREAD

**Discussion by tombelpasso** [micropython/micropython#12628] [ESP32]
The boards I'm using LILYGO_TTGO_LORA32 are very IO limited.  Also it is an outdoor application so using the magnet as input doesn't use any pins and was physically easy to implement.  I sense positive and negative values over a certain threshold and treat them as pins for the primitives/Pushbutton asyncio.   So it is like having two physical pin.
I've been using: LILYGO_TTGO_LORA32-20220618-v1.19.1.bin code

 Is there a way to access the hall_sensor value on the lastest 1.2 version?  
Any advice is welcome.  
If anybody cares, I can post my MagnetPin code which makes it act like a physical pin.

Thanks
- Tom

**Comment by tombelpasso** [comment]
Here's the code I used to simulate a physical pin using the ESP32 hall_sensor.
Reposted, I forgot to import the Pushbutton

```
from math import copysign
import time 
import esp32
from primitives import Pushbutton

class Mag:
    def __init__(self, min=4, offset=0, hyst=2):
        self.min = min
        self.offset = offset
        self.hyst = hyst
        self.last = offset
        self.state = False
        self.raw = 0
        self.n_ave = 8
        self.quit = 30
        
    def read(self):
        v = esp32.hall_sensor()
        self.raw = v
        self.last = v+self.offset
        return self.last
 
    def ave(self, n=None):
        if not n:
            n = self.n_ave
        sum = 0
        for i in range(n):
            sum += self.read()
        v = sum//n
        if abs(v) <= self.min:
            v = 0
        return v
    
    def zero(self,n=8):
        self.offset = -self.ave(n) + self.offset
        return self.offset
    
# Make a fake pin object compatible to use with asyncio primitives 
class Magnet_pin:
    def __init__(self, mag, level=10, hyst=3, inverse=True):
        self.mag = mag
        self.level = level
        self.hyst = hyst
        self.inverse = inverse
        self.state = inverse
        
    def value(self): # simulate physical pin with hall sensor
        m = self.mag.ave()
#         print(m)
        if self.state == self.inverse:
            trig = self.level + self.hyst
        else:
            trig = self.level
        if self.level >= 0:
            greater = m > trig
        else:
            greater = m < trig
        self.state = self.inverse ^ greater
        return self.state
        
    def __call__(self):
        return self.value()

#### Usage example ###

# Use Hall sensor as switch if is a switch
mag = Mag()
mag.zero()
# create fake pins
pinN = Magnet_pin(mag, level=10, hyst=6)  # positive magnet field
pinP = Magnet_pin(mag, -18, -6)

# uses the magnetic pins with primitives/Pushbutton
pbN = Pushbutton(pinN, suppress=True)
pbP = Pushbutton(pinP, suppress=True)
...

**Comment by shariltumin** [comment]
According to Espressif ["Mifgration from 4.4 to 5.0"](https://docs.espressif.com/projects/esp-idf/en/v5.0.4/esp32/migration-guides/release-5.x/peripherals.html?highlight=hall_sensor_read)

" API hall_sensor_read on ESP32 has been removed. Hall sensor is no longer supported on ESP32."

So if you need the hall sensor, you are stuck with MicroPython firmware that was compiled with ESPIDF-4.4. My last local copy of MicroPython that has ```esp32_hall_sensor()``` is **v1.20.0-206-g33b403dfb**.

**Reply by tombelpasso** [reply]
Thanks for the point to the last release version that supports the hall_sensor in firmware.  Looks like I can get up to v1.20 at least.
How can I get v1.20.0-206-g33b403dfb ?

I realize that esp32 hall_sensor reading has been removed from the future releases.  But there must be a way of accessing it.
It looks like it's an adc read operation.  

I would prefer not to be stuck on an old release.
